xmodule/modulestore/split_mongo/split.py

```diff
@@ -233,11 +233,13 @@ def _clear_bulk_ops_record(self, course_key):
             raise TypeError(f'{course_key!r} is not a CourseLocator or LibraryLocator')
 
         if course_key.org and get_library_or_course_attribute(course_key) and course_key.run:
-            del self._active_bulk_ops.records[course_key.replace(branch=None, version_guid=None)]
+            if course_key.replace(branch=None, version_guid=None) in self._active_bulk_ops.records:
+                del self._active_bulk_ops.records[course_key.replace(branch=None, version_guid=None)]
         else:
-            del self._active_bulk_ops.records[
-                course_key.replace(org=None, course=None, run=None, branch=None)
-            ]
+            if course_key.replace(org=None, course=None, run=None, branch=None) in self._active_bulk_ops.records:
+                del self._active_bulk_ops.records[
+                    course_key.replace(org=None, course=None, run=None, branch=None)
+                ]
 
```

Comment:
@bradenmacdonald The reason this [test case](https://github.com/openedx/edx-platform/blob/7d7a18d28bfc794b168ccb2c7b84e23034ebda7d/cms/djangoapps/contentstore/tests/test_contentstore.py#L1132-L1144) was failing was because for some reason the bulk op record is not there when attempting to clear it as part of [self._clear_bulk_ops_record](https://github.com/openedx/edx-platform/pull/34800/files#diff-ce330aeb39241bcb7cee96b41906443b85005e9f9acb84c02c0d1a751f8025aeR319) call. I didn't just want to add this check, I dug deeper to see what is actually causing the issue, especially since it used to work before without these checks.

Oddly enough, after narrowing it down to this [line](https://github.com/openedx/edx-platform/pull/34800/files#diff-ce330aeb39241bcb7cee96b41906443b85005e9f9acb84c02c0d1a751f8025aeR325), when it gets the bulk ops record it (which it does), but the error then occurs in the clearing step. If I remove this line, and just call  `fn()` right away it works fine. 

Before spending more time on this and digging more into it, I just wanted to make sure if my concerns are in the right place, i.e. there is a deeper issue and I should continue investigating, or is this "fix", i.e. adding these checks is a fine approach?

Reply:
@yusuf-musleh Mmmm, that "fix" looks suspicious to me. I prefer not to change split modulestore behavior any more than is necessary, because it can have unforeseen consequences.

What I suspect may be the issue is actually in `BulkOperationsMixin`. If you look at it, most of the existing methods take a parameter `ignore_case=False`, but I didn't implement that for the new `on_commit_changes_to` method. So I suspect that when `create_child` or `update_item` is being called in this case, it has `ignore_case=True` in its kwargs, but we're not passing that into `on_commit_changes_to`, which results in that method creating a different bulk ops record. Or something like that. Otherwise, I'm not even sure how this change could possibly be affecting that test.

Reply:
@bradenmacdonald Agreed, the issue is likely with the underlying implementation related to `ignore_case`. 

> Otherwise, I'm not even sure how this change could possibly be affecting that test.

Especially since, I tried changing the testcase from `course_number.upper() -> course_number.lower()`, just to see if the test passes (since the test checks that the course fails to be created) and it indeed passed.  That could explain why the test is failing.

However there doesn't seem to be a straightforward way to access or determine `ignore_case` inside the [update_item](https://github.com/openedx/edx-platform/pull/34800/files#diff-e2292d98eab04e5ce19b5c784091e24167aa697427772e83752c41f5310abad6R835) method so it can be passed in to the `on_commit_changed_to`.

So I've been toying with the idea of possibly adding a new `ignore_case` field to the `BulkOperationsMixin` that could then be accessible from `update_item` method, it's not quite working yet as I'm figuring out the best way to initialize it/set the value without not make too many changes. Could that be a valid approach, or we'd rather not such changes here, as it might be over complication?

Reply:
@yusuf-musleh The bug turns out to be a pretty subtle one related to the use of `defaultdict` and the fact that calling `_get_bulk_ops_record` _creates_ a new record. So the code in `on_commit_changes_to` had an unintentional side-effect. You can fix it with this change, without needing to modify any other parts of split modulestore:

```diff
         """
         Call some callback when the currently active bulk operation has saved
         """
-        bulk_ops_record = self._get_bulk_ops_record(course_key)
-        if bulk_ops_record.active:
+        # Check if a bulk op is active. If so, defer fn(); otherwise call it immediately.
+        # Note: calling _get_bulk_ops_record() here and then checking .active can have side-effects in some cases
+        # because it creates an entry in the defaultdict if none exists, so we check if the record is active using
+        # the same code as _clear_bulk_ops_record(), which doesn't modify the defaultdict.
+        # so we check it this way:
+        if course_key and course_key.for_branch(None) in self._active_bulk_ops.records:
+            bulk_ops_record = self._active_bulk_ops.records[course_key.for_branch(None)]
             bulk_ops_record.defer_until_commit(fn)
```

Reply:
@bradenmacdonald That was indeed it! Thanks for all your help on this. Overall this bug has been quite the challenge to figure out, especially with all the subtleties 😅